HidPkg/UsbHidDxe/UsbHidDxe.c

```diff
@@ -0,0 +1,1071 @@
+    // Report Descriptor not yet read
+
+    // Get report descriptor length from Hid Descriptor.
+    if (UsbHidDevice->HidDescriptor == NULL) {
+      ASSERT (UsbHidDevice->HidDescriptor != NULL);
+      return EFI_NOT_FOUND;
+    }
+
+    for (Index = 0; Index < UsbHidDevice->HidDescriptor->NumDescriptors; Index++) {
+      if (UsbHidDevice->HidDescriptor->HidClassDesc[Index].DescriptorType == USB_DESC_TYPE_REPORT) {
+        break;
+      }
+    }
+
+    if (Index == UsbHidDevice->HidDescriptor->NumDescriptors) {
+      return EFI_NOT_FOUND;
+    }
+
+    // Index is set to the Report Descriptor index.
+    DescriptorLength               = UsbHidDevice->HidDescriptor->HidClassDesc[Index].DescriptorLength;
+    UsbHidDevice->ReportDescriptor = AllocateZeroPool (DescriptorLength);
+    if (UsbHidDevice->ReportDescriptor == NULL) {
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    Status = UsbGetReportDescriptor (
+               UsbHidDevice->UsbIo,
+               UsbHidDevice->InterfaceDescriptor.InterfaceNumber,
+               (UINT16)DescriptorLength,
+               (UINT8 *)UsbHidDevice->ReportDescriptor
+               );
+    if (EFI_ERROR (Status)) {
+      ASSERT_EFI_ERROR (Status);
+      return Status;
+    }
+
+    UsbHidDevice->ReportDescriptorLength = DescriptorLength;
+  }
+
+  if (*ReportDescriptorSize < UsbHidDevice->ReportDescriptorLength) {
+    *ReportDescriptorSize = UsbHidDevice->ReportDescriptorLength;
+    return EFI_BUFFER_TOO_SMALL;
+  }
+
+  if (ReportDescriptorBuffer == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  CopyMem (ReportDescriptorBuffer, UsbHidDevice->ReportDescriptor, UsbHidDevice->ReportDescriptorLength);
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Retrieves a single report from the device.
+
+  @param  This                  A pointer to the HidIo Instance
+  @param  ReportId              Specifies which report to return if the device supports multiple input reports.
+                                Set to zero if ReportId is not present.
+  @param  ReportType            Indicates the type of report type to retrieve. 1-Input, 3-Feature.
+  @param  ReportBufferSize      Indicates the size of the provided buffer to receive the report.
+  @param  ReportBuffer          Pointer to the buffer to receive the report.
+
+  @retval EFI_SUCCESS           Report successfully returned.
+  @retval EFI_OUT_OF_RESOURCES  The provided buffer is not large enough to hold the report.
+  @retval EFI_INVALID_PARAMETER Invalid input parameters.
+  @retval Other                 Unexpected error reading report.
+**/
+EFI_STATUS
+EFIAPI
+HidGetReport (
+  IN HID_IO_PROTOCOL  *This,
+  IN UINT8            ReportId,
+  IN HID_REPORT_TYPE  ReportType,
+  IN UINTN            ReportBufferSize,
+  OUT VOID            *ReportBuffer
+  )
+{
+  USB_HID_DEV  *UsbHidDevice;
+
+  if ((This == NULL) || (ReportBufferSize == 0) || (ReportBufferSize > MAX_UINT16) || (ReportBuffer == NULL)) {

```

Comment:
The `MAX_UINT16` requirement would be good to mention in the function description.

Reply:
done